### ISSUE TITLE
Win32 Desktop App \include\firebase\firestore\timestamp.h(221,71): error C2589: '(': illegal token on right side of '::'

### ISSUE THREAD

**Issue by heliam1** [firebase/firebase-cpp-sdk#55]
### Please fill in the following fields:
Pre-built SDK from the [website](https://firebase.google.com/download/cpp) or open-source from this repo: Pre-built
Firebase C++ SDK version: 6.14
Firebase plugins in use (Auth, Database, etc.): Firestore, Auth, App
Additional SDKs you are using (Facebook, AdMob, etc.): No
Platform you are using the C++ SDK on (Mac, Windows, or Linux): Windows
Platform you are targeting (iOS, Android, and/or desktop): Desktop, Win32 App

### Please describe the issue here:
When building Win32 Desktop App
This error msg occurs
\include\firebase\firestore\timestamp.h(221,71): error C2589: '(': illegal token on right side of '::'
This is resolved by using
 #undef max
 #undef min
Just above the function, looks like it is caused by windows.h
https://stackoverflow.com/a/2561377/8599411

*There is a workaround to this issue^*
But the actual prevention looks like an easy fix

### Please answer the following, if applicable:
What's the issue repro rate? 100%

**Comment by wu-hui** [comment]
Hi heliam1,

Thank you for reporting this issue!

Yes, `window.h` is the cause here. But I agree with the stackoverflow answer you linked: `#define NOMINMAX` before including `windows.h` is probably the way to go. 

Or maybe you could even have a header for windows specifically, something like:
```
#define NOMINMAX
// Any other defines, e.g. WIN32_LEAN_AND_MEAN
#include <windows.h>
```

**Comment by var-const** [comment]
@heliam1 We plan to change `timestamp.h` in a future release to make sure it compiles even in this case. However, to add to the answer above, this is a well-known, long-standing issue with `<windows.h>` header: it defines `min` and `max` as macros which breaks any code that uses these names to call functions/name local variables/etc. The cleanest workaround is to define a special macro, `NOMINMAX`, _before_ including `<windows.h>` -- this would make the header _not_ define `min` and `max` macros. This might come up useful because you may run into this problem with other libraries, not just Firestore (any header that uses e.g. `std::min` might break if it's included after including `<windows.h>`).

**Comment by patm1987** [comment]
Hi @heliam1,

I'm just checking in, have the answers here been answered to your satisfaction?

**Comment by var-const** [comment]
This should be fixed in [release 6.15.1](https://firebase.google.com/support/release-notes/cpp-relnotes#version_6151_-_june_29_2020).
